NPX

ntwigs opened the pull request:
If the user does not have `create-react-app` installed, but has npx installed - the user will be asked if `npx should be used`.  :rocket:

ntwigs committed:
If user does not have CRA but has npx, use that to run CRA :tada:

ntwigs committed:
Added npxCreateReactApp to initRepo type

ntwigs committed:
CRA with npx command added to initateRepo

ntwigs committed:
Removed whitespace

alexdriaguine commented:
Awesome, thank you! 

My initial though was like this.

1. Check if user has npx. If no, go to step 2. If yes, go to 3
2. Check if user has `create-react-app` installed. If no, do not proceed. 
2. Ask if user wants to use `create-react-app`. If yes, run `create-react-app`. If no, do not proceed.

Or we could ditch global dependencies, an required the user to have `npx` installed (i.e tell them to download it or say that with a certain version of node/npm, they already have it). In this case, a major version will be bumped, and we can later add more templates (next, nuxt etc)..

Thoughts? 😁 I personally feel like the latter option is the better!

ntwigs commented:
The way you are proposing would make it easier to add more `template things`, like "create-vue-app" if that is a thing :grin: 

Counterproposal.
1. Check for npx. If no npx go to step 2, else go to step 3.
2. Check each available `template thing` on the users' system. (Ex. check if create-react-app is installed.) Then go to step 3.
3. Show list of available template things - whereas one is "No Library" which only creates a readme.

alexdriaguine commented:
That sounds perfect! 😁 

So, if a user has no `npx` or any template-stuf installed, they only get the empty one with a readme?

ntwigs commented:
Something like this!
![image](https://user-images.githubusercontent.com/14088342/46867186-685efc00-ce24-11e8-8839-24e428c82e7b.png)

So the list only shows options that are available to the user. If no supported libraries are installed, only the `"no-library"/"empty"/"readme-only"` option will be displayed! :grin:

alexdriaguine commented:
That looks great! 👍 😁

ntwigs committed:
Removed whitespace and renamed test2 to test

ntwigs committed:
Created function for checking if a user has a library

ntwigs committed:
Created functions for handling Libraries and actions

ntwigs committed:
Made it easier to add more libraries

ntwigs commented:
@alexdriaguine Give it a go! :grin:

ntwigs commented:
1. It checks for npx. If npx go to step 2 else step 3.
2. Shows all the available options to the user.
3. Checks the users computer for all supported libraries and then present the user with the available options.

alexdriaguine commented on src/lib/actions.ts:
```diff
@@ -0,0 +1,80 @@
+ *
+ * Good luck!
+ */
+export enum Libraries {
```
This is good, nice helpful comment aswell!

I was wondering if we could make it even more easier to add new libraries by keeping them in an object, like the one returned from getActions.

It could look something like this
```typescript
{
   'create-react-app': (dir: string, npx: boolean) => execute(dir)('create-react-app) // or npx if npx is true
}
```

And when we check if the user has the library as global dependency (in case npx is not available) we could use the key of the object. 

Thoughts? :grin:

alexdriaguine commented on src/lib/entities.ts:
```diff
@@ -13,9 +13,9 @@ export interface GithubRequestHeaders {
   init: () => Promise<any>
-  createReadme: () => Promise<any>
-  createReactApp: () => Promise<any>
+  actions: any
```
If we do as I suggested for the libraries, the actions could follow the same structure, i.e some kind of type `type Action = (dir, npx) => Promise<SOMETHING>`

we could then type this property :grimacing:

alexdriaguine commented on src/main.ts:
```diff
@@ -62,14 +62,10 @@ function main(name: string): void {
     }
     fs.mkdirSync(dir)
-    const useSSHRemote = yield prompt('Use SSH remote instead of https? y/N')
+    const useSSHRemote = yield prompt('Use SSH remote instead of https? y/N: ')
```
Nice, this lack of whitespace has been bothering me forever :laughing:

alexdriaguine reviewed (approved):
Looks really great! 

I have some thoughts that I have commented, but I think this PR works and looks great!

I'm approving this PR, tell me what you think about my comments. We can either merge it as is or dicuss further :grin: :fireworks:

ntwigs committed:
Made it even easier to add a new library (and added input check)

ntwigs committed:
Removed commented code

ntwigs committed:
Fixed package.json

ntwigs commented:
Did some changes!
* Added input validation
* Changed the process for adding more libraries
***
Did not want to do:
`'create-react-app': (dir: string, npx: boolean) => execute(dir)('create-react-app)`
since you shouldn't have to write `(dir: string, npx: boolean) => execute(dir)` for each added library. If you were to use the `key` as a command, then it would have to be the whole command `create-react-app .` or for npx `npx create-react-app .` which wouldn't look as sexy when using the script.
The new and updated version looks like this
```
return {
    'create-react-app': command({
        withNpx: 'npx create-react-app .',
        withoutNpx: 'create-react-app .'
    })
}
```

Which makes it possible to do something like this:

```
return {
    'The one true choice': command({
        withNpx: 'npx create-react-app . --some-important-npx-flag',
        withoutNpx: 'create-react-app .'
    })
}
```
Might be easier to use :sweat_smile: Thoughts? :tada:

alexdriaguine commented:
I think it looks really good! Nice one to add input validation aswell :grin: 
Thanks, I'll merge it and bump a minor version and publish it later today! Good work! :medal_sports:
